### PR TITLE
Alerting: Set Dashboard and Panel IDs on rule group replacement

### PR DESCRIPTION
**Pull request by alexmobo**
**What is this feature?**

This feature fixes a bug that occurred when trying to update an alert rule group (created in the UI) via provisioning where its alerts were attached to a Dashboard/Panel.

**Why do we need this feature?**

When we tried to update an alert rule group (created in the UI) via provisioning where its alerts were attached to a Dashboard/Panel, the DashboardUID and PanelID fields were removed from the database and the alert was detached from the Dashboard/Panel.

**Who is this feature for?**

Users using provisioning for updating alert rule groups created in the UI.

**Which issue(s) does this PR fix?**:

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/models/alert_rule.go [hunk 1]
```diff
@@ -232,10 +232,9 @@ func (alertRule *AlertRule) Diff(rule *AlertRule, ignore ...string) cmputil.Diff
 	return reporter.Diffs
 }
 
-// SetDashboardAndPanel will set the DashboardUID and PanlID
-// field be doing a lookup in the annotations. Errors when
-// the found annotations are not valid.
-func (alertRule *AlertRule) SetDashboardAndPanel() error {
+// SetDashboardAndPanelFromAnnotations will set the DashboardUID and PanelID field be doing a lookup in the annotations.
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/models/alert_rule.go:235]:
```suggestion
// SetDashboardAndPanelFromAnnotations will set the DashboardUID and PanelID field by doing a lookup in the annotations.
```

File: pkg/services/ngalert/models/alert_rule_test.go [hunk 2]
```diff
@@ -82,6 +83,85 @@ func TestErrStateFromString(t *testing.T) {
...
+		{
+			name:                 "dashboardUID and panelID are present and correct",
+			annotations:          map[string]string{DashboardUIDAnnotation: "cKy7f6Hk", PanelIDAnnotation: "65"},
+			expectedError:        nil,
+			expectedDashboardUID: "cKy7f6Hk",
+			expectedPanelID:      65,
+		},
+	}
+
+	for _, testCase := range testCases {
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/models/alert_rule_test.go:149]: It's quite common to abbreviate this as `tc`. It doesn't make a massive difference, but might save some repetition.

```suggestion
	for _, tc := range testCases {
```

**Review by grobinson-grafana — COMMENTED** [review]
Looks good to me, would be great to get a review from @JohnnyQQQQ and @alexweav.

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 3]
```diff
@@ -203,6 +203,10 @@ func (service *AlertRuleService) ReplaceRuleGroup(ctx context.Context, orgID int
 	rules := make([]*models.AlertRule, len(group.Rules))
 	group = *syncGroupRuleFields(&group, orgID)
 	for i := range group.Rules {
+		err := group.Rules[i].SetDashboardAndPanelFromAnnotations()
+		if err != nil {
+			return err
+		}
 		rules = append(rules, &group.Rules[i])
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/provisioning/alert_rules.go:210]:
```suggestion
		rules = append(rules, &group.Rules[i])
```

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 4]
```diff
@@ -203,6 +203,10 @@ func (service *AlertRuleService) ReplaceRuleGroup(ctx context.Context, orgID int
 	rules := make([]*models.AlertRule, len(group.Rules))
 	group = *syncGroupRuleFields(&group, orgID)
 	for i := range group.Rules {
+		err := group.Rules[i].SetDashboardAndPanelFromAnnotations()
+		if err != nil {
+			return err
+		}
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/provisioning/alert_rules.go:209]:
```suggestion
		if err := group.Rules[i].SetDashboardAndPanelFromAnnotations(); err != nil {
			return err
		}
```

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM, good job on adding tests, one small formatting nit.

**Comment by alexmobo** [comment]
Removed the tag `backport v9.0.x` since the endpoint that has the problem didn't exist in that version.

**Pull request merged**